models/issue_watch.go

```diff
@@ -5,86 +5,133 @@
 package models
 
 import (
+	"fmt"
+
+	"code.gitea.io/gitea/modules/log"
 	"code.gitea.io/gitea/modules/timeutil"
 )
 
+// IssueWatchMode specifies what kind of watch the user has on a issue
+type IssueWatchMode int8
+
+const (
+	// IssueWatchModeNone don't watch
+	IssueWatchModeNone IssueWatchMode = iota // 0
+	// IssueWatchModeNormal watch issue (from other sources)
+	IssueWatchModeNormal // 1
+	// IssueWatchModeDont explicit don't auto-watch
+	IssueWatchModeDont // 2
+	// IssueWatchModeAuto watch issue (from AutoWatchOnChanges)
+	IssueWatchModeAuto // 3
+)
+
 // IssueWatch is connection request for receiving issue notification.
 type IssueWatch struct {
 	ID          int64              `xorm:"pk autoincr"`
 	UserID      int64              `xorm:"UNIQUE(watch) NOT NULL"`
 	IssueID     int64              `xorm:"UNIQUE(watch) NOT NULL"`
-	IsWatching  bool               `xorm:"NOT NULL"`
+	Mode        IssueWatchMode     `xorm:"NOT NULL DEFAULT 1"`
 	CreatedUnix timeutil.TimeStamp `xorm:"created NOT NULL"`
 	UpdatedUnix timeutil.TimeStamp `xorm:"updated NOT NULL"`
 }
 
 // IssueWatchList contains IssueWatch
 type IssueWatchList []*IssueWatch
 
-// CreateOrUpdateIssueWatch set watching for a user and issue
-func CreateOrUpdateIssueWatch(userID, issueID int64, isWatching bool) error {
-	iw, exists, err := getIssueWatch(x, userID, issueID)
+// CreateOrUpdateIssueWatchMode set IssueWatchMode for a user and issue
+func CreateOrUpdateIssueWatchMode(userID, issueID int64, mode IssueWatchMode) error {
+	sess := x.NewSession()
+	defer sess.Close()
+	if err := sess.Begin(); err != nil {
+		return err
+	}
+	err := createOrUpdateIssueWatchMode(sess, userID, issueID, mode)
+	if err != nil {
+		return err
+	}
+	return sess.Commit()
+}
+
+func createOrUpdateIssueWatchMode(e Engine, userID, issueID int64, mode IssueWatchMode) error {
```

Comment:
Continuing @guillep2k theme I suppose I should say there's a race here...

This should be an UPSERT style query which updates/inserts as required and then deletes as the last step if the `mode == IssueWatchModeNone`

Reply:
is it still here since IssueWatchModeNone since gone

Reply:
CreateOrUpdateIssueWatchMode create a session witch is used for createOrUpdateIssueWatchMode so there should not be a race ?

Reply:
Sessions will not help with all kinds of race conditions. Check for some explanations [here](https://vladmihalcea.com/how-to-fix-optimistic-locking-race-conditions-with-pessimistic-locking/) and [here](https://stackoverflow.com/a/26081833/225979). But this is the gist of it:

![image](https://user-images.githubusercontent.com/18600385/74618221-938be680-510f-11ea-8504-ebe60224de7e.png)

(Source: [stack overflow](https://stackoverflow.com/questions/46315232/how-can-i-use-transactions-in-mysql-to-avoid-lost-updates))

Reply:
@guillep2k @zeripath this should fix it:
https://github.com/go-gitea/gitea/pull/9535/files#diff-55867f9a81f00e4465d556398d0f159eR45-R46